ui/src/app/utils/cluster-initializer.tsx

```diff
@@ -0,0 +1,301 @@
+import {notebooksClusterApi} from 'app/services/notebooks-swagger-fetch-clients';
+import {clusterApi} from 'app/services/swagger-fetch-clients';
+import {isAbortError, reportError} from 'app/utils/errors';
+import {Cluster, ClusterStatus} from 'generated/fetch';
+
+// We're only willing to wait 20 minutes total for a cluster to initialize. After that we return
+// a rejected promise no matter what.
+const DEFAULT_OVERALL_TIMEOUT = 1000 * 60 * 20;
+const DEFAULT_INITIAL_POLLING_DELAY = 2000;
+const DEFAULT_MAX_POLLING_DELAY = 15000;
+// By default, we're willing to retry twice on each of the state-modifying API calls, to allow
+// for transient 500s.
+const DEFAULT_MAX_CREATE_COUNT = 2;
+const DEFAULT_MAX_DELETE_COUNT = 2;
+const DEFAULT_MAX_RESUME_COUNT = 2;
+const DEFAULT_MAX_SERVER_ERROR_COUNT = 5;
+
+export class ClusterInitializationFailedError extends Error {
+  public readonly cluster: Cluster;
+
+  constructor(message: string, cluster?: Cluster) {
+    super(message);
+    Object.setPrototypeOf(this, ClusterInitializationFailedError.prototype);
+
+    this.name = 'ClusterInitializationFailedError';
+    this.cluster = cluster;
+  }
+}
+
+export class ExceededActionCountError extends ClusterInitializationFailedError {
+  constructor(message, cluster?: Cluster) {
+    super(message, cluster);
+    Object.setPrototypeOf(this, ExceededActionCountError.prototype);
+
+    this.name = 'ExceededActionCountError';
+  }
+}
+
+export class ExceededErrorCountError extends ClusterInitializationFailedError {
+  constructor(message, cluster?: Cluster) {
+    super(message, cluster);
+    Object.setPrototypeOf(this, ExceededErrorCountError.prototype);
+
+    this.name = 'ExceededErrorCountError';
+  }
+}
+
+export interface ClusterInitializerOptions {
+  // Core options. Most callers should provide these.
+  //
+  // The workspace namespace to initialize a cluster for.
+  workspaceNamespace: string;
+  // Callback which is called every time the cluster updates its status. This callback is *not*
+  // called when no cluster is found.
+  onStatusUpdate?: (ClusterStatus) => void;
+  // An optional abort signal which allows the caller to abort the initialization process, including
+  // cancelling any outstanding Ajax requests.
+  abortSignal?: AbortSignal;
+
+  // Override options. These options all have sensible defaults, but may be overridden for testing
+  // or special scenarios (such as an initialization flow which should not take any actions).
+  initialPollingDelay?: number;
+  maxPollingDelay?: number;
+  overallTimeout?: number;
+  maxCreateCount?: number;
+  maxDeleteCount?: number;
+  maxResumeCount?: number;
+  maxServerErrorCount?: number;
+}
+
+/**
+ * A controller class implementing client-side logic to initialize a Leonardo cluster. This class
```

Comment:
question: How much of this class is cluster-specific versus something that could be reused for another long-running operation?